File: src/main/java/io/github/tanguygab/tabadditions/shared/features/chat/emojis/EmojiManager.java
package io.github.tanguygab.tabadditions.shared.features.chat.emojis;

import io.github.tanguygab.tabadditions.shared.features.chat.Chat;
import io.github.tanguygab.tabadditions.shared.features.chat.ChatManager;
import io.github.tanguygab.tabadditions.shared.features.chat.ChatUtils;
import lombok.Getter;
import me.neznamy.tab.api.placeholder.PlaceholderManager;
import me.neznamy.tab.shared.platform.TabPlayer;

import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.regex.Pattern;

public class EmojiManager extends ChatManager {

    private final String output;
    private final boolean untranslate;
    @Getter private final boolean autoCompleteEnabled;
    @Getter private final Map<String,EmojiCategory> emojiCategories = new HashMap<>();
    @Getter private int totalEmojiCount;
    private final Map<TabPlayer, List<String>> emojisAutoCompleteList = new HashMap<>();
    @Getter private final boolean emojisCmdEnabled;

    @SuppressWarnings("unchecked")
    public EmojiManager(Chat chat, String emojiOutput, boolean untranslateEmojis, boolean autoCompleteEmojis, Map<String,Map<String,Object>> emojis, boolean emojisCmdEnabled, boolean toggleCmd) {
        super(chat,toggleCmd,"emojis-off","toggleemojis","chat-emojis");
        setToggleCmdMsgs(translation.emojisOn,translation.emojisOff);
        this.output = emojiOutput;
        this.untranslate = untranslateEmojis;
        autoCompleteEnabled = plugin.getPlatform().supportsChatSuggestions() && autoCompleteEmojis;
        for (String category : emojis.keySet()) {
            Map<String,String> emojisMap = (Map<String,String>) emojis.get(category).get("list");
            totalEmojiCount+=emojisMap.size();
            emojiCategories.put(category,new EmojiCategory(category, emojisMap,ChatUtils.componentToMM((Map<String, Object>) emojis.get(category).get("output"))));
        }

        this.emojisCmdEnabled = emojisCmdEnabled;
        if (emojisCmdEnabled) plugin.getPlatform().registerCommand("emojis");

        if (autoCompleteEmojis)
            for (TabPlayer p : tab.getOnlinePlayers())
                if (!hasCmdToggled(p))
                    loadAutoComplete(p);

        tab.getPlaceholderManager().registerServerPlaceholder("%chat-emoji-total%",-1, ()->String.valueOf(totalEmojiCount));
        chat.placeholders.add(tab.getPlaceholderManager().registerPlayerPlaceholder("%chat-emoji-owned%",5000,p->String.valueOf(ownedEmojis((TabPlayer) p))));
    }

    @Override
    public void unload() {
        if (autoCompleteEnabled) for (TabPlayer p : tab.getOnlinePlayers()) unloadAutoComplete(p);
        super.unload();
    }

    public String process(TabPlayer sender, TabPlayer viewer, String msg) {
        if (hasCmdToggled(viewer) || hasCmdToggled(sender)) return msg;

        for (EmojiCategory category : emojiCategories.values()) {
            Map<String, String> emojis = category.getEmojis();
            if (emojis == null || emojis.isEmpty()) continue;

            for (String emoji : emojis.keySet()) {
                int count = ChatUtils.countMatches(msg, emoji);
                if (count == 0 || emoji.isEmpty()) continue;
                if (!category.canUse(sender,emoji)) {
                    if (untranslate && msg.contains(emojis.get(emoji)))
                        msg = msg.replace(emojis.get(emoji), emoji);
                    continue;
                }
                List<String> list = Arrays.asList(msg.split(Pattern.quote(emoji)));
                msg = "";
                int counted = 0;
                String output = plugin.parsePlaceholders(getOutput(category)
                        .replace("%emojiraw%",emoji)
                        .replace("%emoji%",emojis.get(emoji).replace("\"","''")),
                        sender,viewer);
                if (list.isEmpty()) return output.repeat(count);
                StringBuilder msgBuilder = new StringBuilder(msg);
                for (String part : list) {
                    if (list.indexOf(part)+1 != list.size() || counted != count) {
                        msgBuilder.append(part).append(output);
                        counted++;
                    } else msgBuilder.append(part);
                }
                msg = msgBuilder.toString();
            }
        }
        return msg;
    }

    public String getOutput(EmojiCategory category) {
        return category == null || category.getOutput().isEmpty() ? output : category.getOutput();
    }

    public int ownedEmojis(TabPlayer p) {
        int amt=0;
        for (EmojiCategory category : emojiCategories.values())
            amt+=category.ownedEmojis(p);
        return amt;
    }

    public void loadAutoComplete(TabPlayer p) {
        List<String> list = new ArrayList<>();
        for (EmojiCategory category : emojiCategories.values()) {
            if (category.canUse(p)) {
                list.addAll(category.getEmojis().keySet());
                continue;
            }
            for (String emoji : category.getEmojis().keySet())
                if (category.canUse(p,emoji))
                    list.add(emoji);
        }
        emojisAutoCompleteList.put(p,list);
        plugin.getPlatform().updateChatComplete(p,list,true);
    }
    public void unloadAutoComplete(TabPlayer p) {
        if (emojisAutoCompleteList.containsKey(p))
            plugin.getPlatform().updateChatComplete(p, emojisAutoCompleteList.get(p),false);
    }

    @Override
    public boolean onCommand(TabPlayer sender, String command) {
        if (command.startsWith("/emojis") && emojisCmdEnabled) {
            String cat = command.contains(" ") ? command.split(" ")[1] : "";
            if (cat.isEmpty()) {
                getEmojisCategories(sender);
                return true;
            }

            EmojiCategory category = emojiCategories.get(cat);
            if (category == null || !category.canUse(sender)) {
                sender.sendMessage(translation.emojiCategoryNotFound,true);
                return true;
            }
            getEmojiCategory(sender,category);
            return true;
        }
        return command.equals("/toggleemojis") && toggleCmd(sender);
    }

    private void getEmojisCategories(TabPlayer sender) {
        StringBuilder builder = new StringBuilder();
        AtomicInteger i = new AtomicInteger();
        AtomicInteger emojisOwned = new AtomicInteger();
        emojiCategories.forEach(((categoryName, category) -> {
            if (!category.canUse(sender)) return;
            int owned = category.ownedEmojis(sender);
            if (owned == 0) return;
            emojisOwned.addAndGet(owned);
            builder.append("\n <click:run_command:\"/emojis ").append(categoryName).append("\">")
                    .append(translation.getEmojiCategory(sender, category)).append("</click>");
            i.getAndIncrement();
        }));

        String output = plugin.parsePlaceholders(translation.getEmojiCategoryHeader(i.get(), emojisOwned.get(),emojiCategories.size())+builder,sender);
        chat.sendMessage(sender,chat.mm.deserialize(ChatUtils.toMMColors(output)));
    }

    private void getEmojiCategory(TabPlayer sender, EmojiCategory category) {
        StringBuilder builder = new StringBuilder();
        Map<String,String> emojis = category.getEmojis();
        AtomicInteger i = new AtomicInteger(0);
        emojis.forEach((emoji,output)->{
            if (!category.canUse(sender,emoji)) return;
            builder.append("\n <click:suggest_command:\"").append(emoji).append("\"><insert:\"")
                    .append(emoji).append("\">").append(translation.getEmoji(emoji, output)).append("</insert></click>");
            i.getAndIncrement();
        });

        if (i.get() == 0) {
            sender.sendMessage(translation.emojiCategoryNotFound, true);
            return;
        }
        String output = plugin.parsePlaceholders(translation.getEmojiHeader(i.get(),emojis.size())+builder,sender);
        chat.sendMessage(sender,chat.mm.deserialize(ChatUtils.toMMColors(output)));
    }

}

File: src/main/kotlin/io/github/tanguygab/tabadditions/shared/features/chat/emojis/EmojiManager.kt
package io.github.tanguygab.tabadditions.shared.features.chat.emojis

import io.github.tanguygab.tabadditions.shared.features.chat.Chat
import io.github.tanguygab.tabadditions.shared.features.chat.ChatManager
import io.github.tanguygab.tabadditions.shared.features.chat.ChatUtils.componentToMM
import io.github.tanguygab.tabadditions.shared.features.chat.ChatUtils.countMatches
import io.github.tanguygab.tabadditions.shared.features.chat.ChatUtils.toMMColors
import me.neznamy.tab.shared.config.file.ConfigurationSection
import me.neznamy.tab.shared.platform.TabPlayer

class EmojiManager(
    chat: Chat,
    private val output: String,
    private val untranslate: Boolean,
    autoCompleteEmojis: Boolean,
    emojis: ConfigurationSection,
    emojisCmdEnabled: Boolean,
    toggleCmd: Boolean
) : ChatManager(chat, toggleCmd, "emojis-off", "toggleemojis", "chat-emojis") {
    val autoCompleteEnabled = plugin.platform.supportsChatSuggestions() && autoCompleteEmojis
    private val emojiCategories = mutableMapOf<String, EmojiCategory>()
    private val emojisAutoCompleteList = mutableMapOf<TabPlayer, List<String>>()

    init {
        setToggleCmdMsgs(translation.emojisOn, translation.emojisOff)
        var totalEmojiCount = 0
        for (key in emojis.keys) {
            val category = key.toString()
            val categorySection = emojis.getConfigurationSection(category)
            val emojisMap = categorySection.getMap<String, String>("list", mapOf<String, String>())
            totalEmojiCount += emojisMap.size
            emojiCategories[category] = EmojiCategory(category, emojisMap, componentToMM(categorySection.getConfigurationSection("output")))
        }

        if (emojisCmdEnabled) tab.platform.registerCustomCommand("emojis") { sender, args ->
            if (args.isEmpty()) {
                getEmojisCategories(sender)
                return@registerCustomCommand
            }

            val category = emojiCategories[args.joinToString(" ")]
            if (category == null || !category.canUse(sender)) {
                sender.sendMessage(translation.emojiCategoryNotFound)
                return@registerCustomCommand
            }
            getEmojiCategory(sender, category)
        }
        if (autoCompleteEnabled) tab.onlinePlayers.forEach { if (!hasCmdToggled(it)) loadAutoComplete(it) }

        tab.placeholderManager.registerServerPlaceholder("%chat-emoji-total%", -1) { totalEmojiCount.toString() }
        chat.placeholders.add(
            tab.placeholderManager.registerPlayerPlaceholder("%chat-emoji-owned%", 5000)
            { ownedEmojis(it as TabPlayer).toString() }
        )
    }

    override fun unload() {
        if (autoCompleteEnabled) tab.onlinePlayers.forEach { unloadAutoComplete(it) }
        super.unload()
    }

    fun process(sender: TabPlayer, viewer: TabPlayer?, msg: String): String {
        if (hasCmdToggled(viewer) || hasCmdToggled(sender)) return msg
        var msg = msg

        for (category in emojiCategories.values) {
            val emojis = category.emojis
            if (emojis.isEmpty()) continue

            for ((rawEmoji, emoji) in emojis) {
                val count = countMatches(msg, rawEmoji)
                if (count == 0 || rawEmoji.isEmpty()) continue
                if (!category.canUse(sender, rawEmoji)) {
                    if (untranslate && emoji in msg) msg = msg.replace(emoji, rawEmoji)
                    continue
                }
                val output = plugin.parsePlaceholders(
                    getOutput(category)
                        .replace("%emojiraw%", rawEmoji)
                        .replace("%emoji%", emoji.replace("\"", "''")),
                    sender, viewer
                )
                msg = msg.replace(rawEmoji, output)
            }
        }
        return msg
    }

    private fun getOutput(category: EmojiCategory) = category.output.ifEmpty { output }

    fun ownedEmojis(player: TabPlayer) = emojiCategories.values.sumOf { it.ownedEmojis(player) }

    fun loadAutoComplete(player: TabPlayer) {
        val list = mutableListOf<String>()

        emojiCategories.values.forEach {
            list.addAll(
                if (it.canUse(player)) it.emojis.keys
                else it.emojis.keys.filter { emoji -> it.canUse(player, emoji) }
            )
        }
        emojisAutoCompleteList[player] = list
        plugin.platform.updateChatComplete(player, list, true)
    }

    fun unloadAutoComplete(player: TabPlayer) {
        if (player in emojisAutoCompleteList) plugin.platform.updateChatComplete(player, emojisAutoCompleteList[player]!!, false)
    }

    private fun getEmojisCategories(sender: TabPlayer) {
        val builder = StringBuilder()
        var i = 0
        var emojisOwned = 0
        emojiCategories.forEach { (categoryName: String, category: EmojiCategory) ->
            if (!category.canUse(sender)) return@forEach
            val owned = category.ownedEmojis(sender)
            if (owned == 0) return@forEach
            emojisOwned += owned
            builder
                .append("\n <click:run_command:\"/emojis $categoryName\">")
                .append(translation.getEmojiCategory(sender, category))
                .append("</click>")
            i++
        }

        val output = plugin.parsePlaceholders(
            translation.getEmojiCategoryHeader(i, emojisOwned, emojiCategories.size) + builder,
            sender
        )
        chat.sendMessage(sender, chat.mm.deserialize(toMMColors(output)))
    }

    private fun getEmojiCategory(sender: TabPlayer, category: EmojiCategory) {
        val builder = StringBuilder()
        val emojis = category.emojis
        var i = 0
        emojis.forEach { (emoji: String, output: String) ->
            if (!category.canUse(sender, emoji)) return@forEach
            builder
                .append("\n <click:suggest_command:\"$emoji\"><insert:\"$emoji\">")
                .append(translation.getEmoji(emoji, output))
                .append("</insert></click>")
            i++
        }

        if (i == 0) {
            sender.sendMessage(translation.emojiCategoryNotFound)
            return
        }
        val output = plugin.parsePlaceholders(translation.getEmojiHeader(i, emojis.size) + builder, sender)
        chat.sendMessage(sender, chat.mm.deserialize(toMMColors(output)))
    }

    override fun onCommand(sender: TabPlayer, command: String) {
        if (toggleCmd(sender)) loadAutoComplete(sender)
        else unloadAutoComplete(sender)
    }
}